src/handlers.rs

```diff
@@ -440,56 +442,85 @@ fn handle_make_temp_dir(state: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
   }()))
 }
 
+// This is just type conversion. Implement From trait?
+// See https://github.com/tokio-rs/tokio/blob/ffd73a64e7ec497622b7f939e38017afe7124dc4/tokio-fs/src/lib.rs#L76-L85
+fn convert_blocking<F>(f: F) -> Poll<Buf, DenoError>
+where
+  F: FnOnce() -> DenoResult<Buf>,
+{
+  use futures::Async::*;
+  match tokio_threadpool::blocking(f) {
+    Ok(Ready(Ok(v))) => Ok(v.into()),
+    Ok(Ready(Err(err))) => Err(err),
+    Ok(NotReady) => Ok(NotReady),
+    Err(_) => panic!("blocking error"),
+  }
+}
+
+// TODO Do not use macro for the blocking function.. We should instead be able
+// to do this with a normal function, but there seems to some type system
+// issues. The type of this function should be something like this:
+//   fn blocking<F>(is_sync: bool, f: F) -> Box<Op>
+//   where F: FnOnce() -> DenoResult<Buf>
+macro_rules! blocking {
+  ($is_sync:expr,$fn:expr) => {
+    if $is_sync {
+      // If synchronous, execute the function immediately on the main thread.
```

Comment:
The issue with this is, the blocking function can never have a dependency on an asynchronous function.

Random example:

```
writeFile("somefile", "hello"); <-- async
readFileSync("somefile"); <-- this should wait for writeFile to finish.
```

Even if you don't agree with the above semantics, I predict there will be cases where you do agree.
So I think this is not the right direction -- it'd be better to do block_on(future) instead.

Reply:
The idea with the sync fs functions is that we want to enable people to issue simple syscalls and bipass the complex event machine. Sometimes you want to just block everything and run "mkdir". What your suggesting essentially makes the sync calls into a new form of "await" (with questionable semantics), which blocks the event loop. Now sync calls are no longer syscall invocations - they are complex beasts.

I understand this "chaining" idea you're thinking about, but this isn't the place for it. The sync calls are to bipass all complexity and simply execute a syscall immediately on the main thread.